src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs

```diff
@@ -202,8 +320,13 @@ static IEnumerable<Type> GetParentTypes(Type type)
         }
     }
 
+    bool initialized;
+    readonly List<(Type MessageType, IEnumerable<Type> Hierarchy)> preRegisteredMessagesWithHierarchy = [];
+    readonly List<Type> preRegisteredMessageTypes = [];
```

Comment:
Would it be a good idea to use a ConcurrentBag and eliminate the locsk on add?

Reply:
I removed the locking at a later point because I realized those types were never meant to be thread safe anyway.